File: Tests/PalmierProTests/Audio/AudioMeterTests.swift
import Testing
@testable import PalmierPro

@Suite("Audio meter")
struct AudioMeterTests {
    @Test func analyzesStereoAndClampsRanges() {
        let analysis = AudioLevelAnalyzer.analyze(
            left: [0, -0.5, 0.25, -1],
            right: [0.25, -0.25, 0.25, -0.25],
            range: -4..<20
        )
        let empty = AudioLevelAnalyzer.analyze(left: [], right: [], range: 0..<1)

        #expect(abs(analysis.left.peak - 1) < 0.0001)
        #expect(abs(analysis.left.rms - 0.57282) < 0.0001)
        #expect(abs(analysis.right.peak - 0.25) < 0.0001)
        #expect(abs(analysis.right.rms - 0.25) < 0.0001)
        #expect(empty == .silence)
    }

    @Test func holdsPeakAndDecaysLevel() {
        var state = AudioMeterChannelState()
        state.ingest(rms: 0.25, peak: 0.5, at: 10)
        let initial = state.display(at: 10)
        let afterOneSecond = state.display(at: 11)
        let afterTwoSeconds = state.display(at: 12)

        #expect(abs(initial.levelDb - -12.0412) < 0.001)
        #expect(abs(initial.peakDb - -6.0206) < 0.001)
        #expect(abs(afterOneSecond.levelDb - -36.0412) < 0.001)
        #expect(abs(afterOneSecond.peakDb - initial.peakDb) < 0.001)
        #expect(abs(afterTwoSeconds.peakDb - -15.0206) < 0.001)
    }

    @Test func latchesClippingUntilReset() {
        var state = AudioMeterChannelState()
        state.ingest(rms: 0.5, peak: 1.01, at: 0)
        state.ingest(rms: 0, peak: 0, at: 3)
        #expect(state.display(at: 3).clipped)
        state.resetClipping()
        #expect(!state.display(at: 3).clipped)
    }
}

File: platforms/windows/tests/PalmierPro.Core.Tests/Audio/AudioMeterTests.cs
using PalmierPro.Core.Audio;
using Shouldly;
using Xunit;

namespace PalmierPro.Core.Tests.Audio;

// Verbatim numeric port of Tests/PalmierProTests/Audio/AudioMeterTests.swift — same inputs, same
// expected dB values, so a ballistics regression here would also regress on the Mac.
public sealed class AudioMeterTests
{
    [Fact]
    public void HoldsPeakAndDecaysLevel()
    {
        var state = new AudioMeterChannelState();
        state.Ingest(0.5f, 10);
        var initial = state.Display(10);
        var afterOneSecond = state.Display(11);
        var afterTwoSeconds = state.Display(12);

        initial.LevelDb.ShouldBe(-6.0206f, tolerance: 0.001);
        initial.PeakDb.ShouldBe(-6.0206f, tolerance: 0.001);
        afterOneSecond.LevelDb.ShouldBe(-30.0206f, tolerance: 0.001);
        afterOneSecond.PeakDb.ShouldBe(initial.PeakDb, tolerance: 0.001); // still within the 1.5 s hold
        afterTwoSeconds.PeakDb.ShouldBe(-15.0206f, tolerance: 0.001); // hold expired at 11.5 s, decays 18 dB/s for 0.5 s
    }

    [Fact]
    public void LatchesClippingUntilReset()
    {
        var state = new AudioMeterChannelState();
        state.Ingest(1.01f, 0);
        state.Ingest(0f, 3);
        state.Display(3).Clipped.ShouldBeTrue();

        state.ResetClipping();
        state.Display(3).Clipped.ShouldBeFalse();
    }

    [Fact]
    public void HubIngestsBothChannelsAndResetsToFloor()
    {
        var hub = new AudioMeterHub();
        hub.Ingest(new AudioMeterAnalysis(1f, 0.25f), 0);

        var display = hub.Display(0);
        display.Left.LevelDb.ShouldBe(0f, tolerance: 0.001); // 20*log10(1) == 0 dBFS
        display.Right.LevelDb.ShouldBe(AudioMeterChannelState.Decibels(0.25f), tolerance: 0.001);

        hub.Reset();
        var afterReset = hub.Display(0);
        afterReset.Left.LevelDb.ShouldBe(AudioMeterChannelState.FloorDb, tolerance: 0.001);
        afterReset.Right.LevelDb.ShouldBe(AudioMeterChannelState.FloorDb, tolerance: 0.001);
        afterReset.Left.Clipped.ShouldBeFalse();
    }
}
